Deduplicate JavaDeprecatedAttr in ClassfileParser

som-snytt opened the pull request:
Fixes scala/bug#12799

SethTisue commented:
👀 

```
!!    1 - run/t9644.scala                           [output differs]
% diff /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/run/t9644.check /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/run/t9644-run.log--- empty
+++ t9644-run.log
@@ -1,0 +1,42 @@
+scala.MatchError: new java.lang.annotation.Retention(value = RUNTIME) (of class scala.reflect.internal.Trees$Apply)
+	at Test$.delayedEndpoint$Test$1(t9644.scala:14)
+	at Test$delayedInit$body.apply(t9644.scala:5)
+	at scala.Function0.apply$mcV$sp(Function0.scala:42)
```

som-snytt commented:
I will draw the curtain now and carry on, so to speak.

som-snytt commented:
My first edit retained last added annotation, which I changed when I fixed the failing test. It makes sense to prefer the `DeprecatedAttr` to its runtime twin; probably they are identical. Scala annotations maybe be multiple, such as `throws`.

lrytz commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -1913,8 +1913,23 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
+      def loop(annots: List[AnnotationInfo], rest: List[AnnotationInfo]): List[AnnotationInfo] =
+        rest match {
+          case h :: t =>
+            if (h.atp.typeSymbol.initialize.isJavaDefined && h.matches(annot.atp.typeSymbol.initialize)) Nil
```
I think we should not force symbols here, but do that externally when necessary.

I would also narrow the fix down to the specific issue, i.e., skip adding the annotation in the classfile parser if the symbol already has it.

In general, it's valid to have multiple annotations of the same Java annotation type on a symbol: https://docs.oracle.com/javase/tutorial/java/annotations/repeating.html

lrytz replied:
I should actually read pages I link to before posting 🙈

It's valid to repeat annotations, they are emitted in bytecode using a "container annotation". We have `groupRepeatableAnnotations` in scalac to support that (and emit an error if the annotation is not repeatable).

som-snytt replied:
I didn't like forcing and I didn't like the brittle distinction for "java-defined" annotations. I moved closer to, "Scala should just support all the Java behaviors instead of trying to be clever and different."

som-snytt commented:
Reminder to self to check how deprecation is emitted.

som-snytt committed:
Prefer Java Deprecated with since and use it

In particular, do not retain two copies.

Preserve annotation order although not meaningful.

Co-authored-by: Lukas Rytz <lukas.rytz@gmail.com>

som-snytt commented:
This now implements the lrytz suggestion to confine the special handling to class file parser.

som-snytt commented:
Of course the `since` test must be constrained to `javaVersion: 9+`, as Lukas had mentioned on an old ticket somewhere. So the `neg` test is for that check, and the `run` test always runs (so however CI is set up, it must undergo a jdk 8 test).

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1514,6 +1506,19 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
+          val loopy = addIt(annots, t)
+          if (h.atp.typeSymbol.initialize.isJavaDefined && h.matches(annot.atp.typeSymbol.initialize)) loopy
+          else h :: loopy
+        case Nil => annots
```
```suggestion
        case Nil => List(annot)
```
then we don't need `annots`

som-snytt replied:
That was my original version. I tired of editing back and forth, so I made it a fold just to keep my interest. But I decided not to rely on "order" (whether Deprecated attribute is seen before the RuntimeVisible annotation), so for simplicity it probes for deprecation first.

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -857,16 +856,9 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
-          val annots = new ListBuffer[AnnotationInfo]
           numAnnots times {
-            parseAnnotation(u2()).foreach(annots.addOne)
+            parseAnnotation(u2()).foreach(addUniqueAnnotation(sym, _))
```
Should we restrict to `@Deprecated`, avoid forcing existing symbol annotations for other cases?

som-snytt replied:
Indeed, I wondered if only deprecation has this behavior but did not follow up. (Also whether a name-based check suffices.)

som-snytt replied:
Deprecated is the only "predefined attribute" that becomes an AnnotationInfo. The check is not name-based because we have a symbol.

lrytz commented on test/files/run/t9644.scala:
```diff
@@ -9,7 +9,7 @@ object Test extends App {
   import reflect.runtime.universe._
 
-  val List(d, r) = symbolOf[Foo].annotations
+  val List(r, d) = symbolOf[Foo].annotations
```
Is `r, d` the order in the pickle? Where is the order being reversed (compared to the source file)?

som-snytt replied:
This is addAnnotation preserving order, so this is typing order. Somewhere I said we should verify that it wasn't intentional to reverse order for that reason, or whether order is meaningful.

som-snytt replied:
I added a test for separate compilation, demonstrating the same order. Because I mostly use a landscape-mode monitor with a text editor, I didn't notice that in the presence of a scala sig, we don't see other runtime attributes. Also, I see that on the emitting end, we just set a flag, so I assume asm emits the attribute, which is present in the Scala class file in the test. Probably, adding JavaDeprecatedAttr in class file parser was deemed duplicative; mental Edit: not because of Scala deprecated but because of the runtime annotation. The whole point of this exercise is that we can't rely on the runtime annotation (which scalac itself does not emit, but possibly should, so that Java tooling would see since value written in Scala). Further mental note: Reviewing the long history (2014-2020) of trying to handle java annotations, it's hard to see what worked when. I agree with the original lrytz ticket advocating uniformity.

I probably misunderstood the hrhino comment about order, since the commit says compile order is different. Does order mean source order or insertion order? JVMS does not suggest that attributes or annotations have a natural or unnatural order.

Order could matter, such as `@nowarn` interleaved with buggy annotations. (Hypothetically.)

lrytz replied:
https://github.com/scala/scala/blob/v2.13.11/src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala#L307

The root: https://github.com/scala/scala/commit/828377d9c0#diff-557cf0e98a6fedea5b17028c45a6b2fc2c16a00fb829157721142b4d3e04a606R112

Looking at more history, it seems adding an annotation was always prepending, I guess order wasn't considered important (except for the `reverse` added to the pickler at some point), and prepending is cheaper. Not sure, but if we fix `addAnnotation` we should probably remove the reversal from the pickler.

lrytz commented on test/files/neg/t12799.check:
stale check file?

som-snytt commented:
Thanks, I was so consumed with the previous edit that I did not drill down to answer your question.

lrytz commented:
Are we sure that the `DeprecatedATTR` is visited before the anntoation, or should we safe-guard there too? https://github.com/scala/scala/blob/v2.13.11/src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala#L827-L831

Also, for the attribute we also add an annotation to the synthetic companion module symbol, but we don't do so for the annotation. This means the "since" message doesn't show when accessing a static member:

```scala
scala> new A
           ^
       warning: class A is deprecated (since you like it)
val res0: A = A@2c26ba07

scala> A.foo
       ^
       warning: class A is deprecated
```

som-snytt commented:
I noticed on a recent dotty ticket about `@BeanProperty` that Scala treats `@Deprecated` as an ordinary annotation only. But `@deprecated` distributes deprecation attribute to all `var` accessors. I haven't checked for the Scala annotation yet.

som-snytt commented:
Last commit addresses both review comments: There is no reason to assume attribute ordering, so don't add plain deprecated if it already has it. There is no test. When adding the annotation, also check the static module, verified by test.

Did not squash!

lrytz commented on test/files/run/t12799/C.java:
```diff
@@ -0,0 +1,11 @@
+
+package example;
+
+@Deprecated(since="you like it")
```
Didn't take arguments back on Java 8. We can make this test `// javaVersion: 9+`

som-snytt replied:
oh right. You always say that, yet I always forget. What is Java 8 again?

som-snytt replied:
For the record, made everything a single test because Swiss frugality, but I could also split out the last test. In fact, I will do that because jdk 8 is a major platform.

lrytz reviewed (approved):
LGTM otherwise. I think this can be a single commit.

lrytz reviewed (approved):
Thank you !
